src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/WebhookGoCardlessService.java

```diff
@@ -19,18 +28,35 @@
     private final GoCardlessEventService goCardlessService;
     private final GoCardlessPaymentHandler goCardlessPaymentHandler;
     private final GoCardlessMandateHandler goCardlessMandateHandler;
-
+    private final GoCardlessMandateStateUpdater goCardlessMandateStateUpdater;
+    
     @Inject
     public WebhookGoCardlessService(GoCardlessEventService goCardlessService,
                                     GoCardlessPaymentHandler goCardlessPaymentHandler,
-                                    GoCardlessMandateHandler goCardlessMandateHandler) {
+                                    GoCardlessMandateHandler goCardlessMandateHandler,
+                                    GoCardlessMandateStateUpdater goCardlessMandateStateUpdater) {
         this.goCardlessService = goCardlessService;
         this.goCardlessPaymentHandler = goCardlessPaymentHandler;
         this.goCardlessMandateHandler = goCardlessMandateHandler;
+        this.goCardlessMandateStateUpdater = goCardlessMandateStateUpdater;
     }
 
     public void handleEvents(List<GoCardlessEvent> events) {
         events.forEach(goCardlessService::storeEvent);
+
+        Map<GoCardlessResourceType, List<GoCardlessEvent>> eventsGroupedByResourceType = events.stream().collect(groupingBy(GoCardlessEvent::getResourceType));
+
+        List<GoCardlessEvent> eventsThatAffectMandates =  Optional.ofNullable(eventsGroupedByResourceType.get(GoCardlessResourceType.MANDATES))
+                .orElse(Collections.emptyList());
+
+        determineAffectedMandates(eventsThatAffectMandates)
```

Comment:
Wondering to factor our the mandate processing. So the flow of this method becomes...
1. group events by type
2. process mandate events.
3. in future process other types (e.g. payments).
4. log some stuff

Something like...
```
Map<GoCardlessResourceType, List<GoCardlessEvent>> eventsGroupedByResourceType = events.stream().collect(groupingBy(GoCardlessEvent::getResourceType));
updateStatusForMandateEvents(eventsGroupedByResourceType.getOrDefault(GoCardlessResourceType.MANDATES, Collections.emptyList()));
```
```
private void updateStatusForMandateEvents(List<GoCardlessEvent> eventsThatAffectMandates) {
        determineAffectedMandates(eventsThatAffectMandates)
                .forEach(goCardlessMandateIdAndOrganisationId ->
                        goCardlessMandateStateUpdater.updateState(
                                goCardlessMandateIdAndOrganisationId.getGoCardlessMandateId(),
                                goCardlessMandateIdAndOrganisationId.getGoCardlessOrganisationId()
                        )
                );
    }
```

Reply:
That’s a good idea. I’ll do that.

Reply:
Also can’t believe I used:

```java
Optional.ofNullable(eventsGroupedByResourceType.get(GoCardlessResourceType.MANDATES)).orElse(Collections.emptyList());
```

… when I could have just used:
```java
eventsGroupedByResourceType.getOrDefault(GoCardlessResourceType.MANDATES, Collections.emptyList())
```